--- binutils.orig/binutils/readelf.c	2018-09-05 14:08:22.733186922 +0100
+++ binutils-2.30/binutils/readelf.c	2018-09-05 15:35:38.009749485 +0100
@@ -17634,11 +17634,12 @@ print_gnu_build_attribute_description (E
 
   if (is_open_attr)
     {
+#if 0
       /* FIXME: Need to properly allow for section alignment.  16 is just the alignment used on x86_64.  */
       if (global_end > 0 && start > BFD_ALIGN (global_end, 16))
 	warn (_("Gap in build notes detected from %#lx to %#lx\n"),
 	      global_end + 1, start - 1);
-
+#endif
       printf (_("    Applies to region from %#lx"), start);
       global_offset = start;
 
